Change ActiveModel human_attribute_name to raise an error

shouichi opened the pull request:
### Motivation / Background

When `config.i18n.raise_on_missing_translations = true`, controllers and views raise an error on missing translations. However, models won't. This PR changes models to raise an error when `raise_on_missing_translations` is true.

### Detail

### Additional information

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

MatheusRich reviewed (approved):
I like this. If the action view configuration is on by default, we could make this one too.

MatheusRich commented:
@shouichi can you squash your commits into a single one?

shouichi commented:
@MatheusRich Thank you for the comment. I feel that this PR may contain multiple changes (especially on load hooks). Thus, I opened this PR to gather feedback. I'm OK with squashing or splitting this PR but I want to wait for a wail for other feedback. Thank you!

byroot commented:
Multiple commits can sometimes be justified, but here it's not really the case. Typically if a commit just touch the changelog, it shouldn't be independent.

If I search the commit that introduced a change, I want to be able to blame the CHANGELOG and see the changelog entrie with the actual code change.

So yeah, please squash.

shouichi commented:
@byroot Thank you for the feedback. An option to raise an error on missing translations was added. To implement that feature, a new hook `active_model_translation` was also added. So I wonder if I should split PR or squash commits (sorry for the raw commits though).

byroot commented:
Just squash it makes the most sense here in my opinion

shouichi committed:
Change ActiveModel human_attribute_name to raise an error

When config.i18n.raise_on_missing_translations = true, controllers and
views raise an error on missing translations. However, models won't.
This commit changes models to raise an error when
raise_on_missing_translations is true

Co-authored-by: Alex Ghiculescu <alex@tanda.co>
Co-authored-by: Jean Boussier <jean.boussier@gmail.com>

shouichi commented:
@byroot squashed 👍

ghiculescu commented on activemodel/test/cases/translation_test.rb:
```diff
@@ -126,4 +126,14 @@ def test_human_attribute_name_does_not_modify_options
     assert_equal({ default: "Cool gender" }, options)
   end
+
+  def test_raise_on_missing_translations
+    ActiveModel::Translation.raise_on_missing_translations = true
+
+    assert_raises I18n::MissingTranslationData do
+      Person.human_attribute_name("name")
+    end
+  ensure
+    ActiveModel::Translation.raise_on_missing_translations = false
+  end
```
You could simplify this test using `Object#with`.

```suggestion
  require "active_support/core_ext/object/with" # at top of file
  
  def test_raise_on_missing_translations
    ActiveModel::Translation.with(raise_on_missing_translations: true) do
      assert_raises I18n::MissingTranslationData do
        Person.human_attribute_name("name")
      end
    end
  end
```

shouichi replied:
Thank you!

ghiculescu commented on activemodel/test/cases/translation_test.rb:
```diff
@@ -126,4 +126,14 @@ def test_human_attribute_name_does_not_modify_options
     assert_equal({ default: "Cool gender" }, options)
   end
+
+  require "active_support/core_ext/object/with" # at top of file
```
Sorry I meant that this require should be at the top of this file.

shouichi replied:
I committed the change without a close look :sweat_smile:  I'll fix it.

byroot commented on activemodel/lib/active_model/translation.rb:
```diff
@@ -22,6 +22,8 @@ module ActiveModel
   module Translation
     include ActiveModel::Naming
 
+    mattr_accessor :raise_on_missing_translations, default: false
```
```suggestion
    singleton_class.attr_accessor :raise_on_missing_translations
```

We shouldn't use Ruby class variables, they are awful.

shouichi replied:
OK.

Just out of curiosity, why is `mattr_accessor` awful? I searched rails codebase and about 40 places use it. Also isn't it from active support?

byroot replied:
> Just out of curiosity, why is `mattr_accessor` awful?

See: https://github.com/rails/rails/pull/42451. It uses Ruby class variables (`@@`) which are slow and have wonky semantic, as they leak in sub classes.

For your use case a regular instance variable is faster, simpler.

> Also isn't it from active support?

Yes, but I consider it soft deprecated.

shouichi replied:
Thank you! I learned a lot.

byroot commented on activemodel/lib/active_model/translation.rb:
```diff
@@ -62,11 +64,15 @@ def human_attribute_name(attribute, options = {})
       defaults << :"attributes.#{attribute}"
       defaults << options[:default] if options[:default]
-      defaults << MISSING_TRANSLATION
+      defaults << MISSING_TRANSLATION if !Translation.raise_on_missing_translations
```
```suggestion
      defaults << MISSING_TRANSLATION unless Translation.raise_on_missing_translations
```

byroot commented on activemodel/lib/active_model/translation.rb:
```diff
@@ -62,11 +64,15 @@ def human_attribute_name(attribute, options = {})
-      defaults << MISSING_TRANSLATION
+      defaults << MISSING_TRANSLATION if !Translation.raise_on_missing_translations
+
+      raise_on_missing_translations = true if options[:raise].nil? && Translation.raise_on_missing_translations
```
```suggestion
      raise_on_missing_translations = true if options.fetch(:raise, Translation.raise_on_missing_translations)
```

byroot replied:
Actually, you don't even need that, because `options` is splatted later. So you can just remove all that and call `.translate` this way:

```ruby
      translation = I18n.translate(defaults.shift, count: 1, raise: Translation.raise_on_missing_translations, **options, default: defaults)
```

shouichi replied:
Aha! That's true :bulb:

byroot commented on activemodel/lib/active_model/translation.rb:
```diff
@@ -62,11 +64,13 @@ def human_attribute_name(attribute, options = {})
       defaults << :"attributes.#{attribute}"
       defaults << options[:default] if options[:default]
-      defaults << MISSING_TRANSLATION
+      defaults << MISSING_TRANSLATION unless Translation.raise_on_missing_translations
```
Hum, I'm sorry I was wrong. Here you need to consider both `option` and `Translation.raise_on_missing_translations`

```suggestion
      defaults << MISSING_TRANSLATION unless options.fetch(:raise, Translation.raise_on_missing_translations)
```

shouichi replied:
Thank you!

ghiculescu commented:
Now that I've tested this PR against our codebase, I think we should revert, or make a new config for this 😢 

<img width="351" alt="image" src="https://github.com/user-attachments/assets/3d7f3870-d9bd-4257-a3b3-7edde25eabf1">

There's two types of error that are coming up a lot.

1) is errors when attempting to render a validation error https://github.com/rails/rails/blob/2bb82965a6dd27083ff163efb8170ea63b0e4050/activemodel/lib/active_model/error.rb#L53

2) is errors when rendering a form field (eg. `f.label :foo`).

Previously both of these would generate a default value for the attribute name. Now they will raise, unless you have defined a translation for every single attribute in your app.

Notably even the code in https://guides.rubyonrails.org/getting_started.html#using-a-form-builder would raise if you turned the config on, since nowhere in that guide do we tell the developer to add a translation for `en.activerecord.attributes.article.title`.

byroot commented:
> unless you have defined a translation for every single attribute in your app.

Isn't that the idea? If you internationalized your app, then just "humanizing" the internal field name isn't going to cut it.

Or am I missing something?

That said, I can see how this could break apps that previously had `config.i18n.raise_on_missing_translations` on, and we could indeed introduce a second config to allow for a transition period.

shouichi commented:
Another option would be to allow each model to decide whether to raise an error. This should make our life easier in a transition period.

```
class User
  self.raise_on_missing_translations = true
end

class Post
  self.raise_on_missing_translations = false
end
```

byroot commented:
Yeah, could be useful indeed. Make that variable a `class_variable` so it works with inheritance. Still think we might need either a different global config or a different value `config.i18n.raise_on_missing_translation = :strict` so that apps with `7.2` defaults aren't impacted.

shouichi commented:
@ghiculescu @byroot Thank you for the feedback. I'll submit a PR to address the issue in a few days.

ghiculescu commented:
> If you internationalized your app, then just "humanizing" the internal field name isn't going to cut it.

Yes, if you internationalisd your app, then that's right.

But if you read https://guides.rubyonrails.org/configuring.html#config-i18n-raise-on-missing-translations it doesn't suggest (to me) that this is only a useful config for fully internationalised apps. For example, you might be using i18n just to keep your view files tidy.

shouichi commented:
Submitted https://github.com/rails/rails/pull/52487.

shouichi commented:
Also submitted https://github.com/rails/rails/pull/52495.
